### ISSUE TITLE
When a method / field is created through the suggestions, it is created in a random place

### ISSUE THREAD

**Issue by yuridiniz** [Dart-Code/Dart-Code#972]
## Steps to reproduce
1 - Create two .dart files wich one class for each
```
src/
|---classA.dart
|---classB.dart
```

2 - instantiate ClassA in ClassB 
`
var classA = new ClassA();
`

3 - type a nonexistent function in ClassA
`
classA.nonExistMethod();
`

4 - Use the sugestion (CTRL + .) and select "Create method"

Doing this, will be created a method in a random place on classB.dart file

**Comment by DanTup** [comment]
@yuridiniz By "random", do you mean in an invalid place, or just not neatly sorted as you'd like? Can you give an example of where it ends up and where you think it should?

(@bwilkerson is any attempt made to put this in a good location, or just a valid one?)

**Comment by bwilkerson** [comment]
I don't remember, and can't easily test it at the moment. I suspect that it either adds the method to the end of the class or places it where the sort function would have put it. The latter seems likely because if the members are not sorted then it might well end up in a seemingly random spot.

**Comment by DanTup** [comment]
@yuridiniz Let me know if what you're seeing doesn't seem to match this. Eg. if you run `Sort Members` in `ClassA.dart` and then insert, does it get put in the correct position to keep things sorted? If so, I think this is expected. A concrete example would be useful.

**Comment by yuridiniz** [comment]
Thank you for your support.

Later I will send a gif to demonstrate the problem better.
In fact, it's a totally random place, where it should not be.

For example:

classA.dart

```
class ClassA {
    exitentMethod() {
    }
}
```

classB.dart

```
class ClassB {
    ClassB() {
          var classA = new ClassA();
          classA.createNonExistentMethod(); //Using CTRL + . for create this method
    }
}
```

the result will be:

classB.dart

```
class Cla<METHOD WILL BE CREATE HERE>ssB {
    ClassB() {
          var classA = new ClassA();
          classA.createNonExistentMethod(); //Using CTRL + . for create this method
    }
}
```

I can not tell if it is always created in the same place or if there is a pattern for this error. But later I can see those details better.

**Comment by bwilkerson** [comment]
Ouch! That's definitely not right. Thanks in advance for the extra information!

**Comment by yuridiniz** [comment]
Has a pattern.

When ClassA has nothing, the error occurs in the first line of the classB.dart file, the more lines of code it has in the ClassA class, the lower is the method

![df4lmqbpxa](https://user-images.githubusercontent.com/5683101/40945947-5970e5fe-6831-11e8-8c9f-cfdde4d5b4d8.gif)

**Comment by bwilkerson** [comment]
Ah, the edit is being applied to the wrong file. It was likely suppose to be added at the end of ClassA.

@DanTup Could you check to see whether server is specifying the wrong file when sending back the edits or whether VS Code is applying the edits to the wrong file?

**Comment by DanTup** [comment]
Will do; my guess is it'll be Dart Code's issue - I don't think we supported multi-file edits originally, so I suspect this is using the document from where it was invoked.

**Comment by DanTup** [comment]
Ok, found the issue - when there are multiple edits we used new code that handled this properly, but when there was a single edit we used old bogus code (the reason for different code is that for a single file edit we can use snippets to set tabstops etc., but for multiple we can't).

@bwilkerson I did notice while fixing this that on a `SourceChange` we have a `selection` property, but it's not clear which file this corresponds to (since there's also a collection of edits). I'm going to make the assumption that `selection` applies to the file where the action was initiated but I think it's a little ambiguous (eg. if there's only a single edit in another file, does it mean that one?).

**Comment by DanTup** [comment]
Ok, here's a question... When performing this action, we should we leave the user in the file they were in, or jump them to the location where the new method was created? I can see arguments for both in this case; is it something the server should tell us?

For now I'm making it always leave you in the original file (it's easy to hit Go to Def to get to it, though it's also usually fairly easy to navigate "back") but I'm open to ideas.

**Comment by bwilkerson** [comment]
> I did notice while fixing this that on a SourceChange we have a selection property, but it's not clear which file this corresponds to (since there's also a collection of edits).

The selection property is a Position, which contains a file path and offset within the file. I'm not sure what isn't clear.

> When performing this action, we should we leave the user in the file they were in, or jump them to the location where the new method was created?

I too can see arguments for both. On the one hand, it seems likely that users would want to implement the method they just defined. On the other, it seems likely that users would want to finish the implementation of the method where they invoked the new method. A user study would tell us whether one of those is more common, but I think either behavior is fine for now. (And changes can include edits to more than two files, making it even harder to choose which file to jump to.)

I suspect that leaving the user in the same file is the least surprising, and hence the best, choice.

**Comment by DanTup** [comment]
> The selection property is a Position, which contains a file path and offset within the file. I'm not sure what isn't clear.

I guess my head! You're right, it's right there. That addressing the question perfectly! :-)

In that case, my behaviour is going to be based on the selection - I'll just put focus into the file with that selection. The server in theory has better context about what the action was, so it seems like a good plan to control it from there if we did want to control it (since the rules might be different depending on the type of action).

**Comment by DanTup** [comment]
@yuridiniz btw, this shipped in v2.13.1 today. Please shout if you're still having issues.
